e2e/vmss.go

```diff
@@ -111,7 +112,72 @@ func extractLogsFromVM(ctx context.Context, s *Scenario) {
 		} else {
 			s.T.Logf("logs extracted successfully from VM %q", s.Runtime.VMSSName)
 		}
+		err = extractBootDiagnostics(ctx, s)
+		if err != nil {
+			s.T.Logf("failed to extract boot diagnostics from VM: %s", err)
+		}
+	}
+}
+
+func extractBootDiagnostics(ctx context.Context, s *Scenario) error {
+	// Only extract boot diagnostics for Linux VMs
+	if s.VHD.OS == config.OSWindows {
+		return nil
+	}
+
+	pager := config.Azure.VMSSVM.NewListPager(*s.Runtime.Cluster.Model.Properties.NodeResourceGroup, s.Runtime.VMSSName, nil)
+	for pager.More() {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get VMSS instances: %v", err)
+		}
+
+		for _, vmInstance := range page.Value {
+			// Get boot diagnostics data
+			bootDiagResp, err := config.Azure.VMSSVM.RetrieveBootDiagnosticsData(ctx, *s.Runtime.Cluster.Model.Properties.NodeResourceGroup, s.Runtime.VMSSName, *vmInstance.InstanceID, nil)
+			if err != nil {
+				return fmt.Errorf("failed to get boot diagnostics for VM %s: %v", *vmInstance.InstanceID, err)
+			}
+
+			if bootDiagResp.SerialConsoleLogBlobURI == nil {
+				continue
+			}
+			if *bootDiagResp.SerialConsoleLogBlobURI == "" {
+				continue
+			}
+			// Save serial console log if available
+			logFile := fmt.Sprintf("serial-console-vm-%s.log", *vmInstance.InstanceID)
```

Comment:
is this file getting uploaded to ADO artifact automatically ?

Reply:
yes, the whole directory is.
Eg in the artifacts here: https://msazure.visualstudio.com/CloudNativeCompute/_build/results?buildId=131625449&view=artifacts&pathAsName=false&type=publishedArtifacts
one can directly see that the 2404Gen1 kernel is different than 2404Gen2